### PR TITLE
MigrationModel duplicate entry

### LINKED ISSUE #39866: MigrationModel duplicate entry on Recreate Upgrade in Cluster with 2+ nodes
### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

_No response_

### Describe the bug

When upgrading to a newer Keycloak(26.1.3 -> 26.2.3) in recreate and Cluster with more than 1 node, the `DefaultMigrationManager` runs `migrate` in each node. Specifically it runs `            model.setStoredVersion(currentVersion.toString());` in more than 1 node. 
This results in `MigrationModelEntity` table to have more than 1 row on the same version and the migration was ran more than once. 
Also a custom theme was used. This affects the login(welcome/ui) pages UI to not load properly due to 404 in UI static resources fetching. In the requests with the 404, it can be seen that the Id from the `MigrationModelEntity` table is used so when trying to change it to the other id it works(if the request goes to the same node that created this row).
Each node is using a different Id from the `MigrationModelEntity` table in `MigrationModelAdapter.latest`

Maybe this is a missing configuration or something of this sort, but in previous upgrades(to different verisons) of same scenario it wasn't reproduced.

### Version

26.2.3

### Regression

- [x] The issue is a regression

### Expected behavior

`MigrationModelEntity` table to have a single row for each version.
When upgrading with Recreate + Cluster with more than 1 node, should run these migrations once and should be aware to handle this case of race.

### Actual behavior

`MigrationModelEntity` table have more than 1 row for the version you upgraded to.
When upgrading with Recreate + Cluster with more than 1 node, runs these migrations more than once without logic of locking.

### How to Reproduce?

- Set up Keycloak in version 26.1.3 or lower in Cluster with more than 1 node(can use 4 for easier reproduction)
- Start an upgrade of the environment to 26.2.3, Should be Recreate upgrade(add Debug logs for better visibility).
- Once the upgrade is completed, can be seen in `MigrationModelEntity` table that we have more than 1 row for 26.2.3 version
As this is some kind of race condition, can be a bit of flaky reproduction

### Anything else?

Logs from 2 different pods:
```
local-keycloak-56bdb855f8-gc5jh lightrun-local-keycloak 2025-05-19 13:00:41,596 INFO  [org.keycloak.storage.datastore.DefaultMigrationManager] (main) Migrating older model to 26.2.0-local-keycloak-56bdb855f8-mw2r8 local-keycloak 2025-05-19 13:00:37,876 DEBUG [org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory] (main) Calling migrateModel local-keycloak-56bdb855f8-mw2r8 local-keycloak 2025-05-19 13:00:38,580 INFO  [org.keycloak.storage.datastore.DefaultMigrationManager] (main) Migrating older model to 26.2.0
```

### PR DESCRIPTION
**Pull request by mabartos**
Closes: #39866

- Cleanup for duplicated old records in MIGRATION_MODEL table
- Unique constraint on the VERSION col in MIGRATION_MODEL table
- Advanced lock mechanism will be done in a follow-up issues as this might be backported to 26.2.x

I've reproduced the issue before:
Screenshot From 2025-05-27 14-05-52

After these changes, everything should be good:
Screenshot From 2025-05-27 15-43-50

## Testing approach

- Included these changes into distribution and built custom image `mabartos/keycloak:26.3.0`
- Having Deployment config with 5 replicas
- Initial version Keycloak 26.1.3
- Then upgraded to the `mabartos/keycloak/26.3.0`
- Change sets were propagated (cleanup + unique constraint)
- Table contains only 2 records with versions (26.1.3 and 26.3.0)

<details>
<summary>Used Testing Deployment</summary>

```yaml
apiVersion: v1
kind: Service
metadata:
  name: keycloak
  labels:
    app: keycloak
spec:
  ports:
    - protocol: TCP
      port: 8080
      targetPort: http
      name: http
  selector:
    app: keycloak
  type: ClusterIP
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: keycloak
  # Used to
  name: keycloak-discovery
spec:
  selector:
    app: keycloak
  # Allow not-yet-ready Pods to be visible to ensure the forming of a cluster if Pods come up concurrently
  # publishNotReadyAddresses: true
  clusterIP: None
  type: ClusterIP
---
apiVersion: apps/v1
# Use a stateful setup to ensure that for a rolling update Pods are restarted with a rolling strategy one-by-one.
# This prevents losing in-memory information stored redundantly in two Pods.
#kind: StatefulSet
kind: Deployment
metadata:
  name: keycloak
  labels:
    app: keycloak
spec:
  #serviceName: keycloak-discovery
  # Run with one replica to save resources, or with two replicas to allow for rolling updates for configuration changes
  replicas: 5
  strategy:
    type: Recreate
  selector:
    matchLabels:
      app: keycloak
  template:
    metadata:
      labels:
        app: keycloak
    spec:
      containers:
        - name: keycloak
          image: quay.io/keycloak/keycloak:26.1.3
          args: ["start"]
          env:
            - name: KC_BOOTSTRAP_ADMIN_USERNAME
              value: "admin"
            - name: KC_BOOTSTRAP_ADMIN_PASSWORD
              value: "admin"
            # In a production environment, add a TLS certificate to Keycloak to either end-to-end encrypt the traffic between
            # the client or Keycloak, or to encrypt the traffic between your proxy and Keycloak.
            # Respect the proxy headers forwarded by the reverse proxy
            # In a production environment, verify which proxy type you are using, and restrict access to Keycloak
            # from other sources than your proxy if you continue to use proxy headers.
            - name: KC_PROXY_HEADERS
              value: "xforwarded"
            - name: KC_HTTP_ENABLED
              value: "true"
            # In this explorative setup, no strict hostname is set.
            # For production environments, set a hostname for a secure setup.
            - name: KC_HOSTNAME_STRICT
              value: "false"
            - name: KC_HEALTH_ENABLED
              value: "true"
            - name: 'KC_CACHE'
              value: 'ispn'
            # Use the Kubernetes configuration for distributed caches which is based on DNS
            - name: 'KC_CACHE_STACK'
              value: 'kubernetes'
            # Passing the Pod's IP primary address to the JGroups clustering as this is required in IPv6 only setups
            - name: POD_IP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
            # Instruct JGroups which DNS hostname to use to discover other Keycloak nodes
            # Needs to be unique for each Keycloak cluster
            - name: JAVA_OPTS_APPEND
              value: '-Djgroups.dns.query="keycloak-discovery" -Djgroups.bind.address=$(POD_IP)'
            - name: 'KC_DB_URL_DATABASE'
              value: 'keycloak'
            - name: 'KC_DB_URL_HOST'
              value: 'postgres'
            - name: 'KC_DB'
              value: 'postgres'
            # In a production environment, use a secret to store username and password to the database
            - name: 'KC_DB_PASSWORD'
              value: 'keycloak'
            - name: 'KC_DB_USERNAME'
              value: 'keycloak'
          ports:
            - name: http
              containerPort: 8080
            - name: management
              containerPort: 9000  
          startupProbe:
            httpGet:
              path: /health/started
              port: 9000
            failureThreshold: 15
          readinessProbe:
            httpGet:
              path: /health/ready
              port: 9000
          livenessProbe:
            httpGet:
              path: /health/live
              port: 9000
          resources:
            limits:
              cpu: 2000m
              memory: 4000Mi
            requests:
              cpu: 1500m
              memory: 3500Mi
---
# This is deployment of PostgreSQL with an ephemeral storage for testing: Once the Pod stops, the data is lost.
# For a production setup, replace it with a database setup that persists your data.
apiVersion: apps/v1
kind: Deployment
metadata:
  name: postgres
  labels:
    app: postgres
spec:
  
  replicas: 1
  selector:
    matchLabels:
      app: postgres
  template:
    metadata:
      labels:
        app: postgres
    spec:
      containers:
        - name: postgres
          image: mirror.gcr.io/postgres:17
          env:
            - name: POSTGRES_USER
              value: "keycloak"
            - name: POSTGRES_PASSWORD
              value: "keycloak"
            - name: POSTGRES_DB
              value: "keycloak"
            - name: POSTGRES_LOG_STATEMENT
              value: "all"
          ports:
            - name: postgres
              containerPort: 5432
          volumeMounts:
            # Using volume mount for PostgreSQL's data folder as it is otherwise not writable
            - name: postgres-data
              mountPath: /var/lib/postgresql
      volumes:
        - name: postgres-data
          emptyDir: {}
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: postgres
  name: postgres
spec:
  selector:
    app: postgres
  ports:
    - protocol: TCP
      port: 5432
      targetPort: 5432
  type: ClusterIP
```
</details>

We might probably even update the MigrationModelAdapter to only update fields instead of trying to create a new record.

### CODE REVIEW

**Comment by ahus1** [comment]
> We might probably even update the MigrationModelAdapter to only update fields instead of trying to create a new record.

I disagree with this approach as we want to do an insert, as the insert and the unique constraint prevent doing a migration twice. Doing a migration twice on a realm might lead to wrong results as a developer would have never thought of that situation.

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    protected void generateStatementsImpl() throws CustomChangeException {
+        Set<String> idsToDelete = new HashSet<>();
+
+        String tableName = getTableName("MIGRATION_MODEL");
+        String colId = database.correctObjectName("ID", Column.class);
+
+        try (PreparedStatement ps = connection.prepareStatement(
+                "SELECT ID FROM MIGRATION_MODEL m1 WHERE EXISTS (SELECT 1 FROM MIGRATION_MODEL m2 WHERE m2.VERSION = m1.VERSION AND m2.UPDATE_TIME > m1.UPDATE_TIME)"
```
[the remark this answers is not in the document]
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java:24]: yes, and the example you had above shows that this is possible, so we could end up in a situation where the migration to create an index would fail. 

Instead, you can go for the ìd` column which a string, but guaranteed to be unique.

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            }
+        } catch (Exception e) {
+            throw new CustomChangeException(getTaskId() + ": Failed to detect duplicate MIGRATION_MODEL rows", e);
+        }
+
+        AtomicInteger i = new AtomicInteger();
+        idsToDelete.stream()
+                .collect(Collectors.groupingByConcurrent(id -> i.getAndIncrement() / 20, Collectors.toList())) // Split into chunks of at most 20 items
+                .values().stream()
+                .map(ids -> new DeleteStatement(null, null, tableName)
```
Comment (ahus1) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java:38]: Is it safe to assume that Liquibase will fill the schema and catalog name with proper defaults if we pass null here? I see that we pass null in other places already.

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mabartos) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java]: > non-numbered column IDs

@ahus1 What do you mean exactly?

[the remark this answers is not in the document]
Comment (mabartos) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java]: > Still, this should be (another) follow-up issue.

@ahus1 Agree that we could have the unique constraint on the time. But, does it need to be done as a follow-up? Maybe it'd be good to do it right away, right?

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java (hunk 2 above)
Comment (mabartos) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java]: As you mentioned, we use it on other places, so I assume it worked. But I rather explicitly provided it :)

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mabartos) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java]: I added the unique constraint even on the UPDATE_TIME as we don't expect that 2 different versions can have the same update_time, right?

[the remark this answers is not in the document]
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java]: > I added the unique constraint even on the UPDATE_TIME as we don't expect that 2 different versions can have the same update_time, right?

I wouldn't expect it, but I've seen a lot of things that I didn't expect. And it would be quite difficult to fix for a user if it fails in their production system. So when you choose to add the constraint, then I ask you to add a cleanup script as well. As I mentioned before, I'm ok to leave that for a separate issue/PR.

**Comment by ahus1** [comment]
@mabartos - I reviewed this changed locally and tested some variations: 

* When adding a DB Schema, the delete failed as it was duplicating the schema prefix - and if you pass in the schema, you must not add a dot as a suffix, as Liquibase adds it. The safest way is to pass null, and then Liquibase does the right thing. 
* When there are entries in the table with the same version and the same update_time, the select will not find them, and therefore not delete them. I updated the SQL to compare the IDs instead.
* I see that you pass in a normalized ID column. If we do that, we should do the same for the other columns as well. 

I've pushed a commit that fixed it for me when testing it locally. Please let me know if you agree with these changes.

**Comment by mabartos** [comment]
@ahus1 Thanks for the review and the additional testing!

> The safest way is to pass null, and then Liquibase does the right thing.

+1

> When there are entries in the table with the same version and the same update_time, the select will not find them, and therefore not delete them. I updated the SQL to compare the IDs instead.

Yes, that is true and your changes kinda resolve this issue. However, as IDs of MigrationModel entity (as PKs) are not auto-incremented and referenced as the resource tags, they are randomly generated and cannot be compared based on time of creation. It can lead to a situation in which a more recent record in a table is deleted even it was added later than the one before for the same version. I've slightly edited the SQL script to consider first the update time and then if those are equal choosing some random one. WDYT?

> I see that you pass in a normalized ID column. If we do that, we should do the same for the other columns as well.

+1


Tried the same testing approach and it should work as expected.

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_6_RemoveDuplicateMigrationModelVersion.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    @Override
+    protected void generateStatementsImpl() throws CustomChangeException {
+        Set<String> idsToDelete = new HashSet<>();
+
+        final String tableName = getTableName(MIGRATION_MODEL_TABLE);
+        final String colId = database.correctObjectName("ID", Column.class);
+        final String colVersion = database.correctObjectName("VERSION", Column.class);
+        final String colUpdateTime = database.correctObjectName("UPDATE_TIME", Column.class);
+
+        database.getConcatSql();
```
Comment (mabartos) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_6_RemoveDuplicateMigrationModelVersion.java]: @ahus1 Why is this change necessary?

Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_6_RemoveDuplicateMigrationModelVersion.java]: This was a left-over from my side that should have been removed.

**Comment by mabartos** [comment]
@ahus1 Just an info that I've added another commit to prevent some confusion: https://github.com/keycloak/keycloak/pull/39994/commits/071c5dbe6b0adbfc24d6d6101bf09be0160cb65e

Described in: https://github.com/keycloak/keycloak/pull/39994#issuecomment-2959331690

**Review by ahus1 — APPROVED** [review]
The new SQL is a bit more complicated than I hoped for. I verified it, and also did some manual tests with multiple duplicate entries, and it worked as expected. 

Thank you for this change - approving!

**Comment by mabartos** [comment]
> The new SQL is a bit more complicated than I hoped for.

@ahus1 Yes, me too, but at least we have it semantically correct. I tried to simplify things for reviews as much as possible. 

Thanks!
